File: backend/app/api/global_triggers.py
from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session
from typing import List, Optional
from app.core.database import get_db
from app.models.user import User
from app.models.bot import Bot
from app.models.trigger import Trigger
from app.schemas.trigger import TriggerResponse
from app.utils.dependencies import get_current_user
import logging

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/triggers", tags=["triggers-global"])


@router.get("", response_model=List[TriggerResponse])
async def get_all_triggers(
    bot_id: Optional[int] = Query(None, description="Фильтр по ID бота"),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Получить все триггеры (с опциональным фильтром по боту)"""
    # Получаем все боты пользователя
    user_bots = db.query(Bot).filter(Bot.owner_id == current_user.id).all()
    bot_ids = [bot.id for bot in user_bots]
    
    if not bot_ids:
        return []
    
    # Запрос триггеров
    query = db.query(Trigger).filter(Trigger.bot_id.in_(bot_ids))
    
    # Применяем фильтр по bot_id, если указан
    if bot_id is not None:
        if bot_id not in bot_ids:
            return []  # Бот не принадлежит пользователю
        query = query.filter(Trigger.bot_id == bot_id)
    
    triggers = query.order_by(Trigger.created_at.desc()).all()
    
    # Добавляем информацию о боте к каждому триггеру
    result = []
    for trigger in triggers:
        # Находим бота для этого триггера
        bot = next((b for b in user_bots if b.id == trigger.bot_id), None)
        bot_name = None
        if bot:
            bot_name = bot.name or bot.username or f"Bot #{bot.id}"
        
        result.append(TriggerResponse(
            id=trigger.id,
            bot_id=trigger.bot_id,
            name=trigger.name,
            event_type=trigger.event_type,
            action_type=trigger.action_type,
            action_data=trigger.action_data or {},
            conditions=trigger.conditions or {},
            is_active=trigger.is_active,
            created_at=trigger.created_at,
            updated_at=trigger.updated_at,
            bot_name=bot_name
        ))
    
    return result

File: backend-node/src/api/globalTriggers.ts
import express, { Request, Response } from 'express';
import { authenticateToken } from '../middleware/auth';
import prisma from '../core/database';

const router = express.Router();

/**
 * GET /api/triggers
 * Получить все триггеры (с опциональным фильтром по боту)
 */
router.get('/', authenticateToken, async (req: Request, res: Response) => {
  try {
    const userId = req.user!.userId;
    const botId = req.query.bot_id ? parseInt(req.query.bot_id as string, 10) : undefined;

    // Получаем все боты пользователя
    const userBots = await prisma.bot.findMany({
      where: { ownerId: userId },
    });

    const botIds = userBots.map((bot) => bot.id);

    if (botIds.length === 0) {
      return res.json([]);
    }

    // Запрос триггеров
    const where: any = {
      botId: { in: botIds },
    };

    // Применяем фильтр по bot_id, если указан
    if (botId !== undefined) {
      if (!botIds.includes(botId)) {
        return res.json([]); // Бот не принадлежит пользователю
      }
      where.botId = botId;
    }

    const triggers = await prisma.trigger.findMany({
      where,
      orderBy: { createdAt: 'desc' },
    });

    // Добавляем информацию о боте к каждому триггеру
    const result = triggers.map((trigger) => {
      const bot = userBots.find((b) => b.id === trigger.botId);
      const botName = bot
        ? bot.name || bot.username || `Bot #${bot.id}`
        : null;

      return {
        id: trigger.id,
        bot_id: trigger.botId,
        name: trigger.name,
        event_type: trigger.eventType,
        conditions: trigger.conditions as any,
        actions: trigger.actions as any[],
        is_active: trigger.isActive,
        created_at: trigger.createdAt,
        updated_at: trigger.updatedAt,
        bot_name: botName,
        // Старые поля для обратной совместимости
        action_type: (trigger.actionType as string) || null,
        action_data: (trigger.actionData as any) || null,
      };
    });

    return res.json(result);
  } catch (error) {
    console.error('Error getting all triggers:', error);
    return res.status(500).json({ detail: 'Internal server error' });
  }
});

export default router;
